File: codechef/MayChallenge16/forest.py
N, W, L = map(int, input().split())
heights = [0 for i in range(N)]
rates = [0 for i in range(N)]

for i in range(N):
    heights[i], rates[i] = map(int, input().split())

low = 0
high = 10**18
while low < high:
    mid = low + ((high - low) >> 1)
    total = 0
    exceed = False
    for i in range(N):
        height = heights[i] + rates[i] * mid
        if height >= L:
            total += height
            if total >= W:
                exceed = True
                break
    if exceed:
        high = mid
    else:
        low = mid + 1

print(low)

File: codechef/MayChallenge16/forest.cpp
#include <iostream>

using namespace std;
typedef unsigned long long uint64;
int N;
uint64 W, L;
uint64 heights[100000];
uint64 rates[100000];

int main()
{
	ios_base::sync_with_stdio(0);
	cin >> N >> W >> L;
	for (int i = 0; i < N; i++)
		cin >> heights[i] >> rates[i];
	uint64 low = 0, high = 1e18, mid;
	while (low < high) {
		mid = low + ((high - low) >> 1);
		uint total = 0, previous;
		for (int i = 0; i < N; i++) {
			uint64 height = heights[i] + mid * rates[i];
			if (height < heights[i] || height < mid * rates[i]) {
				total = 1e18;
				break;
			} else if (height >= L) {
				previous = total + height;
				if (previous < total) {
					total = 1e18;
					break;
				}
				total = previous;
			}
		}

		if (total >= W)
			high = mid;
		else
			low = mid + 1;
	}
	cout << low << endl;
}